add ldap group type like posixGroupType

chrismeyersfsu opened the pull request:
related to #557 

* Adds pattern to easy add django-auth-ldap group types classes and to
pass parameters via AUTH_LDAP_GROUP_TYPE_PARAMS
* Adds new group type PosixUIDGroupType that accepts the attribute,
ldap_group_user_attr, on which to search for the user(s) in the group.

chrismeyersfsu committed:
add ldap group type like posixGroupType

* Adds pattern to easy add django-auth-ldap group types classes and to
pass parameters via AUTH_LDAP_GROUP_TYPE_PARAMS
* Adds new group type PosixUIDGroupType that accepts the attribute,
ldap_group_user_attr, on which to search for the user(s) in the group.

cchurch commented on awx/sso/conf.py:
```diff
@@ -295,6 +295,27 @@ def _register_ldap(append=None):
+        category=_('LDAP'),
+        category_slug='ldap',
+        default=collections.OrderedDict([
+            #('member_attr', 'member'),
```
Any reason to leave this commented vs. removing it?

cchurch commented on awx/sso/ldap_group_types.py:
```diff
@@ -0,0 +1,76 @@
+# Copyright (c) 2017 Ansible by Red Hat
```
2018?

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +339,48 @@ def __init__(self, choices=None, **kwargs):
+            params.setdefault('ldap_group_user_attr', 'uid')
+            params_sanitized['ldap_group_user_attr'] = params['ldap_group_user_attr']
+
+        return cls(**params_sanitized)
```
You give `name_attr` as an example parameter than can be configured, but it doesn't appear you're ever passing it to the `LDAPGroupType` class?

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +339,48 @@ def __init__(self, choices=None, **kwargs):
+
+
+class LDAPGroupTypeParamsField(fields.DictField):
+    pass
```
Should there be validation here on which keys are supported?

chrismeyersfsu committed:
more parameters

chrismeyersfsu committed:
validate group type params

wwitzel3 commented on awx/settings/defaults.py:
```diff
@@ -355,6 +355,40 @@ def IS_TESTING(argv=None):
+
+
+generate_ldap_backend(_AUTH_LDAP_SETTINGS_BASE)
+for i in xrange(1, 5):
```
I don't understand this? Why 5?

chrismeyersfsu replied:
arbitrarily chosen.  I'd guess the timeout gods would smite the login process if it exceeded 5.

wwitzel3 reviewed (changes requested):
Lets make this not have import side effects. We can wrap this and call it from the top of our custom _registerr maybe?

chrismeyersfsu committed:
remove uneeded auth ldap settings

* I had thought that setting the settings was required. But carefully
selected defaults for the settings is the correct way to deal with
errors I was seeing early in developing this feature.

wwitzel3 reviewed (approved):
LGTM

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +370,66 @@ def __init__(self, choices=None, **kwargs):
+            return value
+        group_type_str = self.get_depends_on()
+        group_type_str = group_type_str or ''
+        invalid_keys = (set(value.keys()) - set(VALID_GROUP_TYPE_PARAMS_MAP.get(group_type_str, 'LDAPGroupType')))
```
You're validating against the exact group type name here, but using `.endswith()` above. So a group type of `NestedMemberDNGroupType` would validate above, but not here.

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +370,66 @@ def __init__(self, choices=None, **kwargs):
+        # MemberDNGroupType was the only group type, of the underlying lib, that
+        # took a parameter.
+        params_sanitized = dict()
+        if isinstance(cls, LDAPGroupType):
```
This should be `issubclass` instead of `isinstance` -- however I don't think there's a need to check in this case, since every group type is expected to already be a subclass of `LDAPGroupType`.

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +370,66 @@ def __init__(self, choices=None, **kwargs):
+        # took a parameter.
+        params_sanitized = dict()
+        if isinstance(cls, LDAPGroupType):
+            for k in VALID_GROUP_TYPE_PARAMS_MAP['LDAPGroupType']:
```
I would think all of this could be simplified by using `inspect.getargspec(cls.__init__).args[1:]` to determine which arguments are supported for the given class.

chrismeyersfsu replied:
Was worried about kwargs.

cchurch replied:
None of the built-in group types accept unspecified kwargs. If you modified `PosixUIDGroupType` to only accept `name_attr` and `ldap_group_user_attr`, then it would work.

cchurch commented on awx/sso/fields.py:
```diff
@@ -335,19 +370,66 @@ def __init__(self, choices=None, **kwargs):
+        if not value:
+            return value
+        group_type_str = self.get_depends_on()
+        group_type_str = group_type_str or ''
```
To use the `inspect.getargspec(C.__init__).args[1:]` approach here, would need to `find_class_in_modules(group_type_str)`.

chrismeyersfsu committed:
introspect ldap group types for param validation

* Instead of keeping a hard-coded mapping of valid args for each ldap
group type; introspect the subclass to determine valid/invalid fields

cchurch commented on awx/sso/ldap_group_types.py:
```diff
@@ -13,7 +13,7 @@
 
-    def __init__(self, ldap_group_user_attr, *args, **kwargs):
+    def __init__(self, name_attr='cn', ldap_group_user_attr='uid', *args, **kwargs):
         super(PosixUIDGroupType, self).__init__(*args, **kwargs)
```
This won't pass `name_attr` to the superclass.

chrismeyersfsu replied:
indeed

cchurch reviewed (approved):
![image](https://media0.giphy.com/media/CvZuv5m5cKl8c/giphy.gif)
